feat: Add geography support

jimfulton opened the pull request:
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [x] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-bigquery-sqlalchemy/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [x] Ensure the tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)
- [x] Appropriate docs were updated (if necessary)

Fixes #136 🦕

Commit:
Rename refs in files

Commit:
Move package

Commit:
more renaming

Commit:
Expose dialect and types

Commit:
explain the weird registration

Commit:
fixed: No longer need `test_pybigquery_sqla` schema, we generte random schema names.

Commit:
document that we obsolete pybigquery

Commit:
get _type_map from the base module.

It, itself isn't public.

Commit:
fix import

Commit:
Warn if an old (non-empty) version of pybigquery is installed.

Commit:
Add __version__

Commit:
fixed names

Commit:
Updated the way the path to __init__.py is computed

and updated package name in copyright statement.

Commit:
Updated copyright statements with new package name

Commit:
Define the version in version.py

Commit:
skip coverage for code handling pybigquery hanging around

Commit:
blacken/lint

Commit:
Merge branch 'master' into sqlalchemy-bigquery-197

Commit:
merge upstream

Commit:
Merge remote-tracking branch 'origin/sqlalchemy-bigquery-197' into sqlalchemy-bigquery-197

Commit:
merge upstream

Commit:
Start at documentation.

Commit:
system test that reproduces the GeoAlchemy2 core totorial.

Commit:
Added the geography extra

Commit:
Install geoalchemy2 and shapely for system tests when running Python 3.9

Commit:
Initial geography support

Commit:
Added more examples and an installation note.

jimfulton committed:
Update .repo-metadata.json

Co-authored-by: Tim Swast <swast@google.com>

Commit:
added wkb conversion prop and improved the wkr prop to return an object

Commit:
Merge branch 'master' into sqlalchemy-bigquery-197

Commit:
Merge branch 'master' into sqlalchemy-bigquery-197

Commit:
Handle ST_GEOGFROMTEXT calls for bound geoography and GEOGRAPHY type declarations

Commit:
Adjust imports and naming to reflect that we're not using custo classes anymore

Commit:
Added ORM test

Commit:
MOAR tests

and also blacken/lint

Commit:
blacken/lint and don't get Geography from .geography

Commit:
fixed type map, so introspection works.

Commit:
Arrange that string arguments are coerced to geography when calling spacial functions

Commit:
blacken

Commit:
We don't need to explicitly convert strings to geography using WKT

Commit:
Update to reflect automatic string coersion

Commit:
Added missing copyright header

Commit:
removed leftover from moving back to pybigquery

Commit:
blacken

Commit:
import WKY from geography

Mainly so geography is imported, so it's SQLAlchemy compiler tweaks happen.

Commit:
Added missing copyright

Commit:
Go back to inheriting from geoalchemy2 classes

Commit:
Go back to using inherited types

and RST fixes

Commit:
Fixed RST heading level

Commit:
Tie in geography doc.

Commit:
typo

Commit:
need geoalchemy2 for unit tests too

Commit:
comment tweak

Commit:
unit test version of the core system test

Commit:
added missing shapely imports and use WKB's wkt property in bind processor

Commit:
Actually test inserting WKTs

Commit:
Actually test inserting WKTs -- and more edge-case tests

Commit:
Add last_query fixture to make it easier to look at recent queries

Commit:
add a test for st_... functions that don't take geographies and leverage last_query

Commit:
blacken/lint

Commit:
Merge branch 'master' into geography2

Commit:
cleanup "extra" tables to limit test interaction

Commit:
use lake as geography table name

because we don't have to worry about name clashes between tests

Commit:
blacken

Commit:
merge upstream

jimfulton commented on docs/geography.rst:
```diff
@@ -0,0 +1,85 @@
+Geography data is typically represented in Python as text strings in
+WKT format or as `WKB` objects, which contain binary data in WKB
+format.  Querying geographic data returns `WKB` objects and `WKB`
+objects may be used in queries.  When
+calling spatial functions that expect geographic arguments, text
+arguments are automatically coerced to geography.
```
It's true because I made it so. :)  I'm arranging for the coercion to happen at the SQLAlchemy level. See https://github.com/jimfulton/python-bigquery-sqlalchemy/blob/geography2/pybigquery/geography.py#L98-L108

jimfulton commented on noxfile.py:
```diff
@@ -98,6 +98,17 @@ def install_alembic_for_python_38(session, constraints_path):
         session.install("alembic", "-c", constraints_path)
 
 
+def install_geoalchemy_for_python_39(session, constraints_path):
```
OK, I'll try that. I'm not sure what it's going to do with compliance tests.

jimfulton replied:
Done.  Unfortunately, as expected, this didn't work for compliance tests.

jimfulton commented on pybigquery/geography.py:
```diff
@@ -0,0 +1,160 @@
+import sqlalchemy.ext.compiler
+from sqlalchemy.sql.elements import BindParameter
+
+SRID = 4326  # WGS84
```
Sure

jimfulton replied:
done

jimfulton commented on pybigquery/geography.py:
```diff
@@ -0,0 +1,160 @@
+SRID = 4326  # WGS84
+
+
+class WKB(geoalchemy2.WKBElement):
```
OK.

jimfulton replied:
done

jimfulton commented on pybigquery/geography.py:
```diff
@@ -0,0 +1,160 @@
+        super().__init__(data, srid, True)
+
+    @property
+    def wkt(self):
```
OK

jimfulton replied:
done

jimfulton commented on pybigquery/geography.py:
```diff
@@ -0,0 +1,160 @@
+        return WKB(wkb.dumps(wkt.loads(self.data)))
+
+
+class GEOGRAPHY(geoalchemy2.Geography):
```
OK. Do you want to make it "Geography"?

tswast replied:
I suppose most / all of our other type classes are all caps too. Maybe we document it in some contributing docs?

jimfulton replied:
I suggest adding a comment here: 
https://github.com/googleapis/python-bigquery-sqlalchemy/blob/080eda16933be11b156e85d5a83225f5bb8377dc/pybigquery/sqlalchemy_bigquery.py#L128

Although, I'm not sure what to say. :)

Maybe:

`# By convention, dialect-provided types are spelled with all upper case.`

?

jimfulton replied:
Done

jimfulton commented on setup.py:
```diff
@@ -74,6 +74,9 @@ def readme():
         "future",
     ],
+    extras_require=dict(
+        all=["GeoAlchemy2", "shapely"], geography=["GeoAlchemy2", "shapely"],
```
I don't think so.  If they're using alembic, they're install it themselves. pybigquery of more of a plugin for alembic than the other way around.

jimfulton replied:
Well, we did to leverage the automatic extras machinery.

jimfulton commented on tests/system/test_geography.py:
```diff
@@ -0,0 +1,307 @@
+import pytest
+
+pytestmark = pytest.mark.skipif(
+    sys.version_info[:2] != (3, 9), reason="We install geoalchemy2 only for Python 3.9"
```
No, IMO.  I don't want to hide an import problem.  I know that we install geoalchemy2 for Python 3.9 and I want to fail if the import fails for Python 3.9.

tswast replied:
OK

jimfulton replied:
I changed my mind.   Test code should be independent of nox.

If we fail to run some tests due to nox misconfiguration, we'll catch it with coverage.

tswast commented on pybigquery/geography.py:
```diff
@@ -0,0 +1,160 @@
+
+
+@sqlalchemy.ext.compiler.compiles(geoalchemy2.functions.GenericFunction, "bigquery")
+def _fixup_st_arguments(element, compiler, **kw):
```
Could we get a docstring here? I'd like some more context about how this function is expected to be used.

tswast commented on tests/unit/test_geography.py:
```diff
@@ -0,0 +1,156 @@
+import sys
```
Needs license header.

jimfulton replied:
Oops, fixed

Commit:
Add link to WGS83/SRID definition

Commit:
Refactored docs before including generated api documentation

Commit:
MOAR documentation and refactored so WKB only takes a data argument

Commit:
install geoalchemy2 and shapely for docs so we can import and document geography.py

Commit:
refactored docs for for better navigation and to document the geography api

also removed the old api doc, because we want to drop the old apo

Commit:
added missing licence file

jimfulton commented:
Note that I refactored the docs a bit to make navigation easier and to add api documentation for GEOGRAPHY, WKT and WKB.

I also dropped documentation for the old api for doing fry-run queries, since we decided we were going to drop that api.

Commit:
Try to make owlbot/synth do the needsful for extras

Commit:
say the magic word

Commit:
Updated some tests to reflect movement of responsibility.

Commit:
blacken

Commit:
Updated owlbot to leverage extras config and added geoalchemy2 and shapely to doc deps

Commit:
Added test for coverage

jimfulton commented on noxfile.py:
```diff
@@ -84,7 +84,7 @@ def lint_setup_py(session):
 def install_alembic_for_python_38(session, constraints_path):
     """
-    install alembic for Python 3.8 unit and system tests
+    Install alembic for Python 3.8 unit and system tests
```
Done

Commit:
explain why uppercase

Commit:
simplify with importskip

Commit:
Use importskip rather than skipping based on Python version

Test code should be independent from nox, even at the risk of missing
tests.

Commit:
Set up snippets

Commit:
first wrong stab at README

Commit:
Revert "first wrong stab at README"

This reverts commit 2e1fc9a42ec86ed1e8eada0d6d763da1d4fd12b6.

README generation is broke,
https://github.com/googleapis/synthtool/issues/867

so we'll punt for now.

Commit:
typo

Commit:
Add geopandas example

Commit:
Merge branch 'master' into pandas-samples

Commit:
removed unused sys import

Commit:
Added the missing CRS argument.

Commit:
Merge branch 'pandas-samples' of github.com:jimfulton/python-bigquery-sqlalchemy into pandas-samples

Commit:
Removed sphinx roles that aren't understood bt setuptools.

Commit:
Added START and END, but I don't know what I'm doing. :)

Commit:
merge upstream/master

Commit:
merge upstream/master

Commit:
Keep pybigquery history pypi link

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot

See https://github.com/googleapis/repo-automation-bots/blob/master/packages/owl-bot/README.md

tswast commented on docs/geography.rst:
```diff
@@ -0,0 +1,85 @@
+      __tablename__ = 'lakes'
+
+      name = Column(String)
+      geog = column(GEOGRAPHY)
```
```suggestion
      geog = Column(GEOGRAPHY)
```

jimfulton replied:
That's what I get for having docs that aren't automatically tested.  I'll convert this to use a `snippets.py`, following the `python-bigquery` example.

tswast commented on docs/geography.rst:
```diff
@@ -0,0 +1,85 @@
+
+To get geography support, you need to install `pybigquery`
+with the `geography` extra, or separately install `GeoAlchemy2` and
+`shapely`.
```
Let's include an example of the `pip` command in case folks aren't familiar with the "extra" term.

```suggestion
`shapely`.

.. code-block:: console

    pip install 'pybigquery[geography]'
```

jimfulton replied:
OK, will do.

tswast commented on owlbot.py:
```diff
@@ -27,10 +27,19 @@
 # Add templated files
 # ----------------------------------------------------------------------------
+extras = []
+extras_by_python = {
```
Awesome!

tswast commented on owlbot.py:
```diff
@@ -125,6 +103,7 @@ def install_alembic_for_python_38(session, constraints_path):
 
 s.replace( ["noxfile.py"], old_sessions, new_sessions)
 
+# Maybe we can get rid of this when we don't need pytest-rerunfailures
```
Maybe add a link to query retry issue/pr?

jimfulton replied:
OK

jimfulton replied:
Done

tswast commented on pybigquery/geography.py:
```diff
@@ -0,0 +1,242 @@
+        this because when we decide how to bind, we don't always know
+        what the values will be.
+
+        This is not a user-facing methos.
```
```suggestion
        This is not a user-facing method.
```

tswast reviewed (commented):
Looks good! A few minor comments.

gcf-owl-bot[bot] committed:
🦉 Updates from OwlBot

See https://github.com/googleapis/repo-automation-bots/blob/master/packages/owl-bot/README.md

tswast committed:
Merge branch 'master' into pandas-samples

Commit:
Removed the api module

Commit:
added missing comma

Commit:
we decided not to have a shim, so get rid of the check for it.

Commit:
fixed indentation

Commit:
typo

jimfulton committed:
Update pybigquery/geography.py

Co-authored-by: Tim Swast <swast@google.com>

Commit:
Merge branch 'master' into geography2

jimfulton commented:
I added the "do not merge" label.  We won't merge this until the after the rename branch lands.

jimfulton committed:
Show how to install the extra

Co-authored-by: Tim Swast <swast@google.com>

Commit:
More info on why we won't need pytest-rerunfailures

Commit:
Merge branch 'geography2' of github.com:jimfulton/python-bigquery-sqlalchemy into geography2

Commit:
removed extra rst file

Commit:
Added missing import and added module docstring

Commit:
More docs reorg

Commit:
fixed some heading levels and also highlight that older versions are distributed as pybigquery

Commit:
lint

Commit:
Force unique column labels (shouldn't ne necessary)

Commit:
typo

Commit:
fixed snippet tags

Commit:
merge geograph2

Commit:
rename pybigquery to sqlalchemy-bigquery

Commit:
Merge branch 'pandas-samples' into geography3

tswast reviewed (approved):
I like it!

I wonder if we want to merge this before the rename just to make the docfx session happy?

jimfulton commented:
> I wonder if we want to merge this before the rename just to make the docfx session happy?

I'm working on a branch that combines the rename and geography (and snippets to be used in the geography doc).

I suggest we decide next week whether to wait for Dan to fix docfx or to include geography in alpha 1.

Commit:
Remove samples we don't want people to see :)

Commit:
Fixed spelling of sqlalchemy_bigquery

Commit:
Changed the geometry doc to use snippets

and came up with a better query example

Commit:
Added missing licence file

jimfulton commented:
I merged the rename branch into this. Maybe that was a mistake.

We'll either want to merge the rename branch first, at which point this will be smaller, or we'll merge this.

Commit:
blacken

Commit:
lint

Commit:
lint

jimfulton commented:
Everything has been reviewed here except: https://github.com/googleapis/python-bigquery-sqlalchemy/pull/228/commits/6343bee73490b68b513698d0c9a79a99ceaa8438, which changes the geography documentation to use snippets, which found a number of bugs. :)  I also improved one of the query examples.

I tried to format the query examples to be more readable in the documentation, but black wouldn't let me. :(

Commit:
merge master

Commit:
compute all extras

Commit:
blacken

jimfulton commented:
@tswast I'm going to assume you still approve unless I hear otherwise. :smiley: 

I'm thinking of waiting to merge this until after 1.0.0-a1 is released, unless you want me to include these changes in a1.

tswast reviewed (approved):
Looks good!

I'm okay with waiting to merge this until after the a1 release.

Commit:
use `use_labels` to work around another related flakey test.

Commit:
Merge branch 'master' into geography2
